src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java

```diff
@@ -100,14 +100,20 @@ public CapacityResource(
   @Deprecated(since = "https://issues.redhat.com/browse/ENT-4384")
   @ReportingAccessRequired
   public CapacityReport getCapacityReport(
-      ProductId productId,
+      String productIdValue,
       @NotNull GranularityType granularityType,
       @NotNull OffsetDateTime beginning,
       @NotNull OffsetDateTime ending,
       Integer offset,
       @Min(1) Integer limit,
       ServiceLevelType sla,
       UsageType usage) {
+    ProductId productId;
+    try {
+      productId = ProductId.fromString(productIdValue);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e);
+    }
```

Comment:
I'd like to do these checks with a custom validator.  It reduces the boilerplate and it should be pretty easy to write.  We have several examples already in the `swatch-core/src/main/java/org/candlepin/subscriptions/validator` package.

Reply:
I understand your request, and I looked into doing this just now, and there's one small reason I'm nacking this request: Validators don't let us keep the looked-value, so we'd end up effectively scanning the config twice to do essentially the same validation. I wish that JSR-303 had some sort of generic "Converter" interface for this. It's tempting to implement a JAX-RS converter, but that doesn't fit with out OpenAPI generation specs.